_posts/2016-01-08-learn-this-one-weird-trick-to-protect-your-customers-data.markdown

```diff
@@ -0,0 +1,221 @@
+
+### Safe AWS authentication for KMS
+
+Now you're all set, but you have that pesky AWS secret access key that you need to communicate with KMS. It's turtles all the way down again! How do we get past it? We have three great options:
+
+#### IAM roles
+
+If your app is hosted on AWS infrastructure, you're in the clear. EC2 nodes and Lambda functions can be associated with IAM roles -- a set of AWS permissions that API calls originating from that machine use by default. Set up an IAM role that can use KMS's decrypt function with your production master key, associate that role with your production resources, and boom -- now you don't have to store a secret key or authenticate at all. It all just works. Even if you're using Amazon ECS to manage Docker containers!
+
+#### AWS Cognito
+
+AWS has an authentication service geared toward mobile developers (but still quite useful for web applications) called Cognito. Cognito allows you to authenticate with any of the major authentication providers out of the box, or _even hook into your own, home-rolled authentication solution_. Once Cognito deems you authenticated, it can provide you with a set of temporary AWS access credentials with the permissions you need. So however you're handling service authentication now can be baked right in. If you're _not_ handling it in any way right now, there's one last option...
+
+#### Add it during provisioning
+
+This is the least secure-out-of-the-box solution, so don't set this up without a good security engineer on hand. However, when a new server is provisioned, you can have your configuration tool -- Ansible, Chef, Puppet, Salt, etc. -- copy the credential to the box, as long as that tool is running locally and you're not handing that credential off in plaintext to some third party CD solution. This is not recommended, because any attacker or malware scanning the box can find this unencrypted secret and use it against you. However, they cannot do that without generating an event in AWS's thorough CloudTrail audit logs, so frequent (and preferably automated) review of those logs combined with the understanding that a compromised machine assures compromised secrets is still a better solution than plaintext secrets committed to git.
+  
+## Wrapping up
+
+Cryptex is capable of more than I touched on here, so reading through the [Cryptex Documentation] is encouraged. In particular, if the complexities of Amazon KMS are a turn-off, other key sources are available -- and the consistent API makes it easy to plug other custom ones in.
+
+But if you've gotten this far and I still haven't convinced you how important it is to protect your customers' private data with at least the same level of security you require of their own accounts, let me pose a different question: What do you think your customers would prefer?
```

Comment:
Nope, of the customers' (their) accounts. I'm tying it back to the opening paragraph that points out how secure we force our customers to be-- (burdening them with complex passwords or MFA devices.